src/core/namespace.c

```diff
@@ -1754,6 +1762,20 @@ int setup_namespace(
                         };
                 }
 
+                /* Will be used to add bind mounts at runtime */
+                if (setup_propagate) {
+                        const char *root_dir_incoming = prefix_roota(root_directory, incoming_dir);
+
+                        *(m++) = (MountEntry) {
+                                .source_const = propagate_dir,
+                                .path_const = incoming_dir,
+                                .mode = BIND_MOUNT,
+                                /* If root is already mounted read-only we don't set the read-only flag,
+                                * because remounting incoming as read-only will then fail with EPERM */
+                                .read_only = root_image || path_is_read_only_fs(root_dir_incoming) <= 0,
```

Comment:
still not getting why this would fail? is userns involved here?

can you strace this, what call precisely fails?

i.e. remounting a read-only bind mount read-only again should be a NOP. Unless the mount is owner by another userns...

Reply:
I had attached the full debug log in the previous comment thread here: https://github.com/systemd/systemd/pull/17228#discussion_r547015079

The relevant error is:

```
Dec 22 01:14:48 systemd-testsuite systemd[29]: Failed to re-mount '/run/systemd/unit-root/run/host/incoming' and its submounts: Operation not permitted
```

ie, the remount to make it read-only called from here: https://github.com/systemd/systemd/blob/70d6631bf4b21f00d5bfb4629e3dc13f53ed1cbe/src/core/namespace.c#L1925

Reply:
and yes it's a a bunch of units with user namespaces, eg: systemd-logind

Reply:
(strace is tricky, as it's one of the integratiton tests in qemu that fails)

Reply:
Ok, I've got it sorted, and it was obvious in hindsight (which is always 20/20) - I picked /run/host/incoming as the inner mount directory. When running under unprivileged nspawn, starting a MountAPIVS=yes service without an empty /run results in trying to remount the /run/host/incoming made available by nspawn, which fails with EPERM. It's a nesting problem, so to speak.

Fixed by simply changing the inner directory to /run/systemd/incoming - which matches the outer, /run/systemd/propagate.

PTAL!